bionic/flow.py

```diff
@@ -1167,19 +1167,18 @@ def name(self):
 
         return self.get('core__flow_name')
 
-    def render_dag(self, include_core=False, vertical=False, curvy_lines=False):
+    def render_dag(self, include_core=False, vertical=False, curvy_lines=False, svg_format=False):
```

Comment:
You know, thinking more about this, I wonder if the user should need to specify a format here at all. If they're rendering the output in Jupyter, then we can choose the best output for them, which is SVG -- I don't see much reason someone would want to insist on PNG in Jupyter. And if they want to write to a file, they can provide more information when they call `save(...)`. So maybe it'd be better if we returned our own image class that implements `_repr_svg_` and `save(...)` (which can handle SVG itself and forward other file formats to PIL). What do you think?

If you do decide to have a format parameter where, it might be better to have a string parameter (like `format='svg'`) in case we want to (a) introduce more formats, or more likely (b) transition to making SVG the default, which probably means going through a deprecation phase where you get a warning if you don't explicitly specify a format.

Reply:
For saving, if we allow for SVGs it's currently hard to decouple image rendering and saving. With PIL I think you can take bytes from a PNG and save as (any other supported format) but it looks like SVGs need to be created and saved specifically as SVGs in both stages. So I think the format parameter is required at both steps, not just in `.save()`. 

I tried pseudocoding several designs that might fulfill this. Haven't tested any for real yet, Let me know if I understood the intention correctly! 

1) create new image class that forwards file formats to PIL or SVG: For PIL, use the built in .save() and .show(). For SvgImage, use the methods (.save, ._repr_svg_) specific to that class. Looks kind of redundant but maybe there's some shorthand for this? 
```
class FlowImage:
    def __init__(self, pydot_graph, format=None):
        if format == 'png':
            self.image = Image(BytesIO(pydot_graph.create_png())
	else: # default
            self.image = SvgImage(pydot_graph.create_svg())
	
    def save(fp='foo/bar/baz'):
        self.image.save(fp)
	
    def show # Using any PIL format, but it'll error out on SVG
        self.image.show()
```

2) Having a FlowImage class without the additional SvgImage class, all conditional logic is handled in this class directly
```
class FlowImage:
    def __init__(self, pydot_graph, format='svg'):
        self.format = format
        self.image_data = pydot_graph.create_{format}()
    if self.format == 'svg':
        self.image = self.image_data
    else:
        self.image = Image(BytesIO(self.image_data))

    def show(self):
        if self.format == 'svg':
            # raise some warning
        else:
            self.image.show()
    def _repr_svg_(self):
        return self.image

    def save(self):
        if self.format == 'svg':
            # save self.image_data
        else:
            self.image.save()
```
3) Rewrite the preexisting `image_from_dot` to return either a PIL `Image` or `SvgImage`. This is closest to what we currently have and doesn't require a more general class that encapsulates both PIL and Svg. 
```
def image_from_dot(pydot_graph, format='svg'):
	if format == 'png':
            return Image(BytesIO(pydot_graph.create_png())
	return SvgImage(pydot_graph.create_svg())
```
Between all the above I think #3 is the simplest. What are your thoughts?

Re: Jupyter, it makes sense that SVG is the preferred format! We can introduce a check for the Python platform like [this](https://stackoverflow.com/questions/15411967/how-can-i-check-if-code-is-executed-in-the-ipython-notebook)).

Reply:
I think all of those are decent approaches -- the one I was thinking of was closest to 2. But it's still not clear to me that we need to pass an explicit `format` argument to the constructor, because we can figure out which format to use later depending on what the user asks for.

* `_repr_svg_`: We should use the best format available, which (thanks to you!) is SVG.
* `show`: SVG is not available so we might as well use PNG. If we implement support for SVG later I think we can safely switch to that.
* `save`: I think this is the only place where the user will have a preference, and we can figure out what it is from the arguments they pass in.

The main reason I'd like to avoid taking the `format` argument is so that users immediately get the improved SVG view without having to do anything extra. (You could make the default value `'svg'`, but then any existing code using `flow.render_dag().save('flow.png')` will break.)

What do you think?

Reply:
💡 I think I figured it out